File: src/mleLoop.c
# include <R.h>
# include <Rmath.h>

void mleLoopCoe(double *alpha, double *doobs, double *c1, double *c2, double *beta, double *irr, double *wtr, double *kz, double *dosat, int *nobs){
	int i;
	double a1, z=0; // z has to be a double b/c kz is passed in as a double (for cases where it isn't 0)
	// double alpha=*alpha, doobs=*doobs, c1=*c1, c2=*c2, beta=*beta, irr=*irr, wtr=*wtr, kz=*kz, dosat=*dosat;
	int ni=*nobs;
	for(i=1; i<ni; i++){
		
		if(kz[i-1] != z){ // if k.gas is 0, kz is 0
			
			a1 = *c1 * irr[i - 1] + *c2 * log(wtr[i - 1]) + kz[i - 1] * dosat[i - 1];
			
			alpha[i] = a1 / kz[i - 1] + -exp(-1 * kz[i - 1]) * a1 / kz[i - 1] + beta[i - 1] * alpha[i-1];
			
		}else{ // if kz is 0, must use different equations to avoid /0
			
			alpha[i] = alpha[i-1] + *c1 * irr[i - 1] + *c2 * log(wtr[i - 1]);
			
		}
		
	}
}

void mleLoopCpe(double *alpha, double *doobs, double *c1, double *c2, double *beta, double *irr, double *wtr, double *kz, double *dosat, int *nobs){
	int i;
	double a1, z=0; // z has to be a double b/c kz is passed in as a double (for cases where it isn't 0)
	// double alpha=*alpha, doobs=*doobs, c1=*c1, c2=*c2, beta=*beta, irr=*irr, wtr=*wtr, kz=*kz, dosat=*dosat;
	int ni=*nobs;
	for(i=1; i<ni; i++){
		
		if(kz[i-1] != z){ // if k.gas is 0, kz is 0
			
			a1 = *c1 * irr[i - 1] + *c2 * log(wtr[i - 1]) + kz[i - 1] * dosat[i - 1];
			
			alpha[i] = a1 / kz[i - 1] + -exp(-1 * kz[i - 1]) * a1 / kz[i - 1] + beta[i - 1] * doobs[i-1];
			
		}else{ // if kz is 0, must use different equations to avoid /0
			
			alpha[i] = doobs[i-1] + *c1 * irr[i - 1] + *c2 * log(wtr[i - 1]);
			
		}
		
	}
}

File: R/mleLoop.R
# double *alpha, double *doobs, double *c1, double *c2, double *beta, double *irr, double *wtr, double *kz, double *dosat, int *nobs
mleLoopR <- function(alpha, doobs, c1, c2, beta, irr, wtr, kz, dosat){
	# dyn.load("/Users/Battrd/Documents/School&Work/WiscResearch/LakeMetabolizer/src/LakeMetabolizer.so")
	# dyn.unload("/Users/Battrd/Documents/School&Work/WiscResearch/LakeMetabolizer/src/LakeMetabolizer.so")
	nobs <- length(doobs)
	a.loop <- .C("mleLoopC", alpha=as.double(alpha), as.double(doobs), as.double(c1), as.double(c2), as.double(beta), as.double(irr), as.double(wtr), as.double(kz), as.double(dosat), as.integer(nobs), PACKAGE="LakeMetabolizer")
	# a.loop[["alpha"]]
	return(a.loop[["alpha"]])
}
